subsys/mgmt/mcumgr/transport/src/smp_udp.c

```diff
@@ -231,98 +303,106 @@ static void create_thread(struct config *conf, const char *name)
 	k_thread_start(&(conf->thread));
 }
 
-SYS_INIT(smp_udp_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
-
 int smp_udp_open(void)
 {
-	struct config *conf;
+	bool started = false;
 
 #ifdef CONFIG_MCUMGR_TRANSPORT_UDP_IPV4
-	struct sockaddr_in addr4;
-
-	memset(&addr4, 0, sizeof(addr4));
-	addr4.sin_family = AF_INET;
-	addr4.sin_port = htons(CONFIG_MCUMGR_TRANSPORT_UDP_PORT);
-	addr4.sin_addr.s_addr = htonl(INADDR_ANY);
-
-	conf = &configs.ipv4;
-	conf->sock = create_socket((struct sockaddr *)&addr4, conf->proto);
-
-	if (conf->sock < 0) {
-		return -MGMT_ERR_EUNKNOWN;
+	if (!IS_THREAD_RUNNING(configs.ipv4.thread)) {
+		(void)k_sem_reset(&configs.ipv4.network_ready_sem);
+		create_thread(&configs.ipv4, "smp_udp4");
+		started = true;
+	} else {
+		LOG_ERR("IPv4 UDP MCUmgr thread is already running");
 	}
-
-	create_thread(conf, "smp_udp4");
 #endif
 
 #ifdef CONFIG_MCUMGR_TRANSPORT_UDP_IPV6
-	struct sockaddr_in6 addr6;
-
-	memset(&addr6, 0, sizeof(addr6));
-	addr6.sin6_family = AF_INET6;
-	addr6.sin6_port = htons(CONFIG_MCUMGR_TRANSPORT_UDP_PORT);
-	addr6.sin6_addr = in6addr_any;
-
-	conf = &configs.ipv6;
-	conf->sock = create_socket((struct sockaddr *)&addr6, conf->proto);
-
-	if (conf->sock < 0) {
-		return -MGMT_ERR_EUNKNOWN;
+	if (!IS_THREAD_RUNNING(configs.ipv6.thread)) {
+		(void)k_sem_reset(&configs.ipv6.network_ready_sem);
+		create_thread(&configs.ipv6, "smp_udp6");
+		started = true;
+	} else {
+		LOG_ERR("IPv6 UDP MCUmgr thread is already running");
 	}
-
-	create_thread(conf, "smp_udp6");
 #endif
 
-	return MGMT_ERR_EOK;
+	if (started) {
+		/* One or more threads were started, send interface notifications */
+		conn_mgr_resend_status();
+	}
+
+	return 0;
 }
 
 int smp_udp_close(void)
 {
 #ifdef CONFIG_MCUMGR_TRANSPORT_UDP_IPV4
-	if (configs.ipv4.sock >= 0) {
+	if (IS_THREAD_RUNNING(configs.ipv4.thread)) {
 		k_thread_abort(&(configs.ipv4.thread));
-		close(configs.ipv4.sock);
-		configs.ipv4.sock = -1;
+
+		if (configs.ipv4.sock >= 0) {
+			close(configs.ipv4.sock);
+			configs.ipv4.sock = -1;
+		}
```

Comment:
If the thread is running, `.sock` gets initialised by the thread itself, therefore there is no need to reinitialise it. And if `.sock` < 0, `close()` would safely return -1 and set `errno = EBADF`.
```suggestion
		close(configs.ipv4.sock);
```

Reply:
Sock is not guaranteed to set by the thread, it would only be set if the network interface was or has came up since the thread has started, if not, it would be -1 and it doesn't make sense to try closing an invalid socket.